### ISSUE TITLE
primitives: Consider removing `serde` stuff

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3034]
There is some debate about what benefit/costs serde's `derive` feature gives us. Before releasing `primitives` we should re-visit the topic and make a decision.

Original discussion is in https://github.com/rust-bitcoin/rust-bitcoin/pull/2991 

Note especially: https://github.com/rust-bitcoin/rust-bitcoin/pull/2991#issuecomment-2226025957

**Comment by Kixunil** [comment]
To be clear:

* Keep obvious impls
* Fix or remove non-obvious impls
* Change the remaining derives to manual impls (may use decl macros)

**Comment by apoelstra** [comment]
We should also come up with a game plan for what happens if we remove the "non-obvious impls" like those on `TxOut` or `Transaction` which are ad-hoc and defined as "whatever serde derive does with the struct as we defined it back in 2015", and then people show up saying that they have a critical dependency on this.

I think we can create a `bitcoin-legacy-serde` crate that re-implements these things, provides a module, and then these people can add `#[serde(with = "bitcoin_legacy_serde::transaction")]` etc to their existing stuff, with very little in terms of breakage.

Maybe we should do this on day one, and put this in the CHANGELOG, and tell people that if they *do* use the legacy thing, to maybe file an issue so we can be aware of whatever they're doing.

**Comment by Kixunil** [comment]
> I think we can create a `bitcoin-legacy-serde` crate that re-implements these things, provides a module, and then these people can add `#[serde(with = "bitcoin_legacy_serde::transaction")]` etc to their existing stuff, with very little in terms of breakage.

I like this. `#[serde(remote = "...")] should do the job.

> Maybe we should do this on day one

Yes, I we shouldn't release `primitives` with things we know we want to remove.

**Comment by apoelstra** [comment]
I had never heard of serde(remote) but this is the second time you mentioned it so I looked it up: https://serde.rs/remote-derive.html

Very nifty.

**Comment by apoelstra** [comment]
One somewhat-annoying thing is that `bitcoin-legacy-serde` will need to depend on `primitives`, and we won't be able to release it until `primitives` is released. So we will need to be a bit quick on the draw to do both releases at the same time

**Comment by Kixunil** [comment]
> So we will need to be a bit quick on the draw to do both releases at the same time

Shouldn't be a too big of a deal to run `cargo publish` twice.

**Comment by tcharding** [comment]
I don't know what the current status of this issue is?

**Comment by apoelstra** [comment]
I think we should just drop the serde stuff on `primitives` and wait for complaints. Migrating to 1.0 will be a big deal for people and nobody is going to upgrade with so much urgency that they can't wait for us to write some adaptor structs (or they can write them themselves).

**Comment by apoelstra** [comment]
Also, the status might be "we've done this over the last few weeks". I'm not sure either.

**Comment by tcharding** [comment]
Excuse me for being braindead but I"m still a bit lost here. Does #4595 close this?

**Comment by apoelstra** [comment]
@tcharding I'll review #4595 with this issue in mind. I think the answer is yes.

**Comment by tcharding** [comment]
As shown in #4789 we can serde any type that implements consensus encoding traits. For 1.0 should we remove all serde stuff from any type that implements the traits and pull the `consensus::serde` module into the new `consensus_encoding` crate and be done with it?

**Comment by apoelstra** [comment]
Yeah, for 1.0 that sounds like a plan. Possibly we will want to "bless" certain serde impls by making them the default but that's a non-breaking change that we can do later.

**Comment by tcharding** [comment]
I started working on this and got a bad feeling when removing `serde` derives from simple wrapper types. Downstream Tobin would be annoyed if I had to write a long attribute just to consensus encode a simple wrapper type.

**Comment by apoelstra** [comment]
Ok, nice, that's good feedback. Let's not do it then :).
